internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go

```diff
@@ -0,0 +1,255 @@
+	Cache  *sidecarproxycache.Cache
+}
+
+func (f *Fetcher) FetchWorkload(ctx context.Context, id *pbresource.ID) (*intermediateTypes.Workload, error) {
+	rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
+
+	switch {
+	case status.Code(err) == codes.NotFound:
+		// We also need to make sure to delete the associated proxy from cache.
+		// We are ignoring errors from cache here as this deletion is best effort.
+		f.Cache.DeleteSourceProxy(resource.ReplaceType(types.ProxyStateTemplateType, id))
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+
+	w := &intermediateTypes.Workload{
+		Resource: rsp.Resource,
+	}
+
+	var workload pbcatalog.Workload
+	err = rsp.Resource.Data.UnmarshalTo(&workload)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&workload, err)
+	}
+
+	w.Workload = &workload
+	return w, nil
+}
+
+func (f *Fetcher) FetchProxyStateTemplate(ctx context.Context, id *pbresource.ID) (*intermediateTypes.ProxyStateTemplate, error) {
+	rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
+
+	switch {
+	case status.Code(err) == codes.NotFound:
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+
+	p := &intermediateTypes.ProxyStateTemplate{
+		Resource: rsp.Resource,
+	}
+
+	var tmpl pbmesh.ProxyStateTemplate
+	err = rsp.Resource.Data.UnmarshalTo(&tmpl)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&tmpl, err)
+	}
+
+	p.Tmpl = &tmpl
+	return p, nil
+}
+
+func (f *Fetcher) FetchServiceEndpoints(ctx context.Context, id *pbresource.ID) (*intermediateTypes.ServiceEndpoints, error) {
+	rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
+
+	switch {
+	case status.Code(err) == codes.NotFound:
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+
+	se := &intermediateTypes.ServiceEndpoints{
+		Resource: rsp.Resource,
+	}
+
+	var endpoints pbcatalog.ServiceEndpoints
+	err = rsp.Resource.Data.UnmarshalTo(&endpoints)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&endpoints, err)
+	}
+
+	se.Endpoints = &endpoints
+	return se, nil
+}
+
+func (f *Fetcher) FetchDestinations(ctx context.Context, id *pbresource.ID) (*intermediateTypes.Destinations, error) {
+	rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
+
+	switch {
+	case status.Code(err) == codes.NotFound:
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+
+	u := &intermediateTypes.Destinations{
+		Resource: rsp.Resource,
+	}
+
+	var destinations pbmesh.Upstreams
+	err = rsp.Resource.Data.UnmarshalTo(&destinations)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&destinations, err)
+	}
+
+	u.Destinations = &destinations
+	return u, nil
+}
+
+func (f *Fetcher) FetchDestinationsData(
+	ctx context.Context,
+	destinationRefs []intermediateTypes.CombinedDestinationRef,
+) ([]*intermediateTypes.Destination, map[string]*intermediateTypes.Status, error) {
+
+	var destinations []*intermediateTypes.Destination
+	statuses := make(map[string]*intermediateTypes.Status)
```

Comment:
nit:
One of my personal style things is to avoid serial adjacent type declarations, especially mixing `var` and `:=` in favor of a `var(...)` block instead. Mostly this helps call out "hey here's a bunch of related variables i'm declaring at the same time" which can aid in readability.

```suggestion
	var (
		destinations []*intermediateTypes.Destination
		statuses     = make(map[string]*intermediateTypes.Status)
	)
```